File: 059_SubarrayProductLessThanK.cpp
// Topic: Sliding Window / Two Pointers
// Subarray Product Less Than K: use sliding window maintaining product
// Time Complexity: O(n)
// Space Complexity: O(1)

#include <bits/stdc++.h>
using namespace std;

int numSubarrayProductLessThanK(vector<int>& nums, int k)
{
    if (k <= 1)
        return 0;
    long long prod = 1;
    int left = 0, res = 0;
    for (int right = 0; right < nums.size(); ++right)
    {
        prod *= nums[right];
        while (prod >= k)
        {
            prod /= nums[left++];
        }
        res += right - left + 1;
    }
    return res;
}

int main()
{
    cout << numSubarrayProductLessThanK((vector<int>{10, 5, 2, 6}), 100) << "\n";  // 8
    cout << numSubarrayProductLessThanK((vector<int>{1, 2, 3}), 0) << "\n";        // 0
    return 0;
}

File: 059_SubarrayProductLessThanK.py
"""
Topic: Sliding Window / Two pointers
Problem: Subarray Product Less Than K
Description: Count contiguous subarrays whose product is less than k.
Time: O(n)
Space: O(1)
"""

from typing import List


def numSubarrayProductLessThanK(nums: List[int], k: int) -> int:
    if k <= 1:
        return 0
    prod = 1
    left = 0
    ans = 0
    for right, val in enumerate(nums):
        prod *= val
        while prod >= k:
            prod //= nums[left]
            left += 1
        ans += right - left + 1
    return ans


if __name__ == '__main__':
    print(numSubarrayProductLessThanK([10,5,2,6], 100))  # 8
    print(numSubarrayProductLessThanK([1,2,3], 0))  # 0
